### ISSUE TITLE
no-referrer-when-downgrade - no DB creation, error concerning url

### ISSUE THREAD

**Issue by AlexSchenker** [FelixHenninger/lab.js#21]
Dear Felix, 
thank you very much for providing us with lab.js, I am using it to implement a pilot study on our new University server right now. 

Your explanations are in general clrear to me, also for the PHP changes: https://labjs.readthedocs.io/en/latest/learn/deploy/2-php.html
Though, I am running into a problem now, see below.

Thanks a lot for your feedback and thoughts!
Best wishes,
Alexandra

### What would you like or expect to see?

I find the no-referrer-when-downgrade error in the debug window and it does not create the data/data.sqlite and therefore does not collect the experimental data.

### What is currently happening?

I find the no-referrer-when-downgrade error in the debug window and it does not create the data/data.sqlite and therefore does not collect the experimental data.
-> The directory rights are set to 777 at the point of testing. The .htaccess file is present.
-> We are working with a Url with http:// only, hence no https://

We are working with Apache, PHP is installed and activated.
Though I wonder -> do we have to change over to https:// (SSL encryption) or change something on the Apache WebServer to get that error sorted and are able to collect data?

### Please describe the steps needed to reproduce this phenomenon

It happens each time, no matter what I change with file directories or removing / renaming .htaccess. 

### Finally, please tell us which browser you're using (and the version thereof)

Google Chrome and Firefox

**Comment by FelixHenninger** [comment]
Hej Alexandra,

thanks for checking out `lab.js`, and also for reporting this issue! I'm not sure what's going on here, it looks like you're doing everything exactly right -- I hadn't tried hosting studies via `http` before, but it's working for me at least on my server a moment ago, so that can't be the issue. 🤔 I think we might have to do some sleuthing to figure out what's happening.

I'm especially puzzled by the `no-referrer-when-downgrade` error -- we do store the `referrer` in the library to figure out where people came to the study from, but I always thought that would just be an empty value if something went wrong. I also don't quite get the `downgrade` part -- I thought that could only happen if somebody moved from an encrypted (`https`) to an unencrypted (`http`) page. Do you get more information with the error, and does it occur immediately after you've loaded the study page, or only after you've moved beyond the first page (and the data starts to get transmitted)?

One thing that could be helpful is if you could share a screenshot from the network tab of your browser's debug tools, while the study is running. In particular, it would be great to see if your server sets a `Referrer-Policy` in the response headers (on the right-hand side), which you can look at by selecting any of the requests:

![request headers](https://user-images.githubusercontent.com/8349942/46492555-317a5c00-c80e-11e8-8e15-cc9ead643ec4.png)

Ok, that's all I have for now -- I hope this helps a bit, please let me know if I can support you on the way! I'll also be online in our [Slack channel](http://slackin-nmbrcrnchrs.herokuapp.com/), which you're very welcome to join if you'd like more chatty Q&A.

Either way, thanks again for reporting this, I'm looking forward to figuring this one out! Kind regards,


-Felix

**Comment by AlexSchenker** [comment]
Hi Felix, thanks for your quick reply. This was much faster than I expected! :-)

When I load the study it looks fine, it also runs through OK. In the Debug window, there is a Flavicon Error, though I did not thinkg this would be important. 
Then the 500 errors start showing when I go through the pages where the people are responding and the scripts should record data.

I am creating a few print screens, so maybe you can see what is going on. The IP address is covered up, it is always the same IP.

Curious what you can see... Thanks a lot!
Best wishes,
Alexandra

![1-flavicon ico-404](https://user-images.githubusercontent.com/43858007/46495105-d7c96000-c814-11e8-86c7-ea1671054721.JPG)
![2-internal-server-error-500](https://user-images.githubusercontent.com/43858007/46495125-def06e00-c814-11e8-94e8-0a9c6262f3b0.JPG)
![3-directs to source data js url](https://user-images.githubusercontent.com/43858007/46495135-e44db880-c814-11e8-8e03-1ae8e79c4dcf.JPG)
![4-backend php-headers](https://user-images.githubusercontent.com/43858007/46495140-e748a900-c814-11e8-8906-164bbeaa2d1a.JPG)

**Comment by AlexSchenker** [comment]
Btw: I think this is our Referrer Policy -> no-referrer-when-downgrade error but there is this 500 internal error

**Comment by FelixHenninger** [comment]
Hej Alexandra,

thanks for _your_ super-quick response! (as for myself, I tend to run the late shift -- I try, but I can't promise to always keep up this speed, though! 😉). 

From what you've posted, I think that the issue may be on the server side after all -- the `500` error code indicates, I think, that the server is unhappy. Since the response comes through, I'm guessing that the connection is working ok -- as you've noted, one of the errors comes from a missing favicon; also, the error location in your third screenshot is the point where the data is transmitted (as noted before, we also save the `referrer`, but that happens elsewhere -- if the error come from there, you'd have spotted it). I'm no longer sure that the `no-referrer-when-downgrade` policy is at fault here -- it seems to be the default setting, and also present on our server (should have checked that earlier, sorry!).

That leaves us to look on the server for any issues with the data collection. Do you have access to the server log files? They should usually contain debugging information in case something's going wrong on the server-side. For example, our server generates the following error if it can't write to the database because it's missing permissions:

```
[Thu Oct 04 21:25:23.922925 2018] [:error] [pid 5274] [client 134.155.xxx.yyy:50024] PHP Fatal error:
  Uncaught PDOException: 
    SQLSTATE[HY000] [14] unable to open database file in /srv/www/virtual/jdm/htdocs/stroop-test/backend.php:7
  Stack trace:
    #0 /srv/www/virtual/jdm/htdocs/stroop-test/backend.php(7): PDO->__construct('sqlite:/srv/www...')
    #1 {main}
      thrown in /srv/www/virtual/jdm/htdocs/stroop-test/backend.php on line 7, referer: https://www.jdm.uni-mannheim.de/stroop-test/
```

(you probably know best where your server keeps the logs -- on our's they're in `/var/log/apache2/error.log`)

Ok, I hope this makes sense -- let me know what you find out, and also if there's anything I can do to help!

-Felix

PS: Oh, and just FYI, I'm not sure if it's your server's address, but there is still a `Host` IP address visible in the last screenshot 😉.

**Comment by FelixHenninger** [comment]
(If you can't access the log files, let me know, and I'll work out a version of the backend that provides you with feedback about what's going wrong -- we just don't usually include that debug information for security reasons)

**Comment by AlexSchenker** [comment]
Hi Felix
Thanks a lot for your detailed information! I appreciate it! And I did not expect you to answer this quick ;-)

Myself, I do not have access to the server. My Professor, who I am running the experiment with, will have to check the logs at some point. It is a new server he had setup recently for our faculty. 

If the backend version would be a quick thing, that could be an option which helps us. Else I will just update the ticket as soon as I have the info. 

In the meantime I will work on an option to have a *.csv with the stimuli imported as I got around 240 lines and adding them manually will take me quite a while.
Is there something in lab.js which supports this and I did not see yet? If so, I can open a new ticket and we can discuss this there.

Again a big thank you, after all lab.js is free and I did not expect such great and detailed support!
I used to be a Tech. IT Customer Support specialist so I know how many hours of work are in this.

Best wishes,
Alexandra

PS: Thanks for the feedback concerning IP on the last screenshot, I removed that image.

**Comment by FelixHenninger** [comment]
Hej Alexandra,

thanks for your updates! I've put together a quick `PHP` script that runs through the checks needed to get the database up and running -- if you'd like to give it a try, you'll need to upload it as `test.php` into your study folder, alongside `backend.php`:

```
<!doctype html>
<html>
<head>
  <title>lab.js backend checks</title>
</head>
<body>
<?php
  define('DB_PATH', getcwd() . '/data/data.sqlite');

  echo 'Active PHP version: ' . phpversion() . '<br>';

  $pdo_available = extension_loaded('PDO');
  echo 'PDO database bindings available: ' .
    ($pdo_available ? 'yes' : 'no') .
    '<br>';

  $pdo_sqlite_available = in_array(
    'sqlite',
    PDO::getAvailableDrivers()
  );
  echo 'PDO sqlite binding available: ' .
    ($pdo_sqlite_available ? 'yes' : 'no') .
    '<br>';

  $pdo_sqlite_loaded = extension_loaded('pdo_sqlite');
  echo 'PDO sqlite binding loaded: ' .
    ($pdo_sqlite_loaded ? 'yes' : 'no') .
    '<br>';

  $folder_exists = file_exists('data');
  echo 'Data folder exists: ' .
    ($folder_exists ? 'yes' : 'no') .
    '<br>';

  $folder_writable = is_writable('data');
  echo 'Data folder writable: ' .
    ($folder_writable ? 'yes' : 'no') .
    '<br>';

  if ($folder_writable) {
    $db_exists = file_exists(DB_PATH);
    echo 'Database file exists: ' .
      ($db_exists ? 'yes' : 'no') .
      '<br>';

    $db_writable = is_writable(DB_PATH);
    echo 'Database file writable: ' .
      ($db_writable ? 'yes' : 'no').'<br>';
  }

  $db = NULL;

  echo 'Setting up PDO connection: ';
  try {
    $db = new PDO('sqlite:' . DB_PATH);
  } catch (Exception $e) {
    echo 'exception ' . $e->getMessage() . '<br>';
  } finally {
    if ($db) {
      echo 'success (created database file if not already present)<br>';
    }
  }

  $request_url = 'http' .
    (isset($_SERVER['HTTPS']) && $_SERVER['HTTPS'] === 'on' ? 's' : '') .
    '://' . $_SERVER[HTTP_HOST] . $_SERVER[REQUEST_URI];
  $db_url = str_replace('test.php', 'data/data.sqlite', $request_url);

  $request = curl_init($db_url);
  curl_setopt($request, CURLOPT_RETURNTRANSFER, true);
  curl_setopt($request, CURLOPT_VERBOSE, false);
  $result = curl_exec($request);
  $code = curl_getinfo($request, CURLINFO_HTTP_CODE);

  echo 'Response code trying to access the database from the web' .
    '(should be 40x or possibly 50x): ' . $code . '<br>';
?>
</body>
</html>
```

It should output some diagnostic information if you access it with your browser.

> In the meantime I will work on having a *.csv with the stimuli imported as I got around 240 lines and adding them manually will take me quite a while.
> I just have seen your builder 2017.2.0 which I am downloading now to have a look at this CSV Import Export for loops :-)

Ah, awesome -- did you get the `CSV` import to work? The button in the interface is really well hidden, if you have an idea on how to make that more findable, I'd love to hear it!

> Again a big thank you, after all lab.js is free and I did not expect such great and detailed support!
I used to be a Tech. IT Customer Support specialist so I know how many hours of work are in this.

Oh, thanks, that's super-kind -- I really enjoy helping people with their research, and it's a huge pleasure to hear from everyone, get feedback and see the cool stuff y'all are doing! 🙂 If you'd like to join some more direct Q&A, most of the support happens via our [slack channel](http://slackin-nmbrcrnchrs.herokuapp.com/), which you're super-welcome to join if you like.

Ok, so much for now -- I hope this helps you get a bit closer to finding out what's going on! Have a great weekend, and the very best regards,


-Felix

**Comment by AlexSchenker** [comment]
Dear Felix

Thanks a lot for your kind help! We do have a problem on the server. My Prof also did some further tests and confirmed. At the moment we will work in a different way, until we can fix it.

And thanks for the hint concerning hidden CSV feature. I found it and will start to work with it now. :-)

Best wishes,
Alexandra

**Comment by FelixHenninger** [comment]
Hi Alex,

thanks a lot for your interest, and for helping us work through this! Please don't hesitate to let me know if you have further questions around this or anything else -- and again please also be warmly invited to [join us on Slack](http://slackin-nmbrcrnchrs.herokuapp.com/), where you'll find help yet more quickly.

All the best, and hope to hear from you soon!


-Felix
